packages/office-ui-fabric-react/src/components/Dropdown/Dropdown.base.tsx

```diff
@@ -539,7 +539,43 @@ export class DropdownBase extends React.Component<IDropdownInternalProps, IDropd
   private _onRenderList = (props: ISelectableDroppableTextProps<IDropdown, HTMLDivElement>): JSX.Element => {
     const { onRenderItem = this._onRenderItem } = props;
 
-    return <>{props.options.map((item: any, index: number) => onRenderItem({ ...item, index }, this._onRenderItem))}</>;
+    const groups: IDropdownOption[][] = [[]];
+    let index = -1;
+
+    for (const item of props.options) {
+      const groupLength = groups.length;
+      const currentGroup = groups[groupLength - 1];
+      if (item.itemType === SelectableOptionMenuItemType.Header) {
+        if (currentGroup.length === 0) {
```

Comment:
>currentGroup [](start = 12, length = 12)

I'm not seeing the usage of this except for as a marker for if you are already processing a group (e.g. have seen a header). Does this need to be an array?

Reply:
my initial thought was to transform the data into an array of arrays....one for each group, then render them all out.
```
[
  [header1, item2, item3],
  [header4, item5, item6]
]
```
Then I can loop through each array and render out the group down below. 

I could consider refactoring this to simply be a queue....stack a bunch of items in until i hit a header, then spit them out and start the queue over again. Wasn't sure if that refactoring was adding any value.

Reply:
If you did that (batched all items associated with a header together and flushing them when encountering another header) you should end up having to iterate over the items less, yeah?

---
In reply to: [353948752](https://github.com/OfficeDev/office-ui-fabric-react/pull/11374#discussion_r353948752) [](ancestors = 353948752)

Reply:
hmmm, yeah, instead of pushing the options to the array, i could push the rendered options....then render the header at the end with the associated wrapping parent. I'll give that a shot tomorrow to see if I can reduce the processing.

Reply:
Ok, i got this working with a single forEach loop, and a single temp cache. I'm also breaking after a divider, so items after the divider, but before the next header, will not be in a group.